### ISSUE TITLE
Cannot join a Saved Question with sorted metric

### ISSUE THREAD

**Issue by flamber** [metabase/metabase#13744]
**Describe the bug**
When creating a question, and joining a Saved Question, which is sorted/ordered by a metric, then the query fails with invalid reference to aggregation index.
Workaround is to remove the sorting/ordering from the Saved Question.

**To Reproduce**
1. Custom question > Sample Dataset > Products - summarize "Count" by "Category", sorted by "Count" - save as "Q1"
![image](https://user-images.githubusercontent.com/1447303/98679918-fa69c380-2360-11eb-9230-fc14392e94f2.png)
2. Custom question > Sample Dataset > Products - join with "Q1" on Category=Category
![image](https://user-images.githubusercontent.com/1447303/98680122-4157b900-2361-11eb-99a6-fc2f6f264a0d.png)
3. Visualize - fails with `No aggregation at index: 0`

<details><summary>Full stacktrace</summary>

```
2020-11-10 14:30:55,344 INFO middleware.fetch-source-query :: Fetched source query from Card 351: 
 {:source-table 10, :aggregation [[:count]], :breakout [[:field-id 107]], :order-by [[:asc [:aggregation 0]]]}

2020-11-10 14:30:55,349 INFO middleware.fetch-source-query :: Fetched source query from Card 351: 
 {:source-table 10, :aggregation [[:count]], :breakout [[:field-id 107]], :order-by [[:asc [:aggregation 0]]]}

2020-11-10 14:30:55,682 ERROR middleware.catch-exceptions :: Error processing query: null
{:database_id 4,
 :started_at #t "2020-11-10T14:30:55.339215+01:00[Europe/Copenhagen]",
 :json_query
 {:type "query",
  :query
  {:source-table 10,
   :joins [{:fields "all", :source-table "card__351", :condition ["=" ["field-id" 107] ["joined-field" "Question 351" ["field-literal" "CATEGORY" "type/Text"]]], :alias "Question 351"}]},
  :database 4,
  :parameters [],
  :middleware {:js-int-to-string? true, :add-default-userland-constraints? true}},
 :native nil,
 :status :failed,
 :class java.lang.Exception,
 :stacktrace
 ["--> mbql.util$fn__27019$aggregation_at_index__27028$fn__27031.invoke(util.clj:541)"
  "mbql.util$fn__27019$aggregation_at_index__27028.invoke(util.clj:530)"
  "driver.sql.query_processor$fn__68790.invokeStatic(query_processor.clj:472)"
  "driver.sql.query_processor$fn__68790.invoke(query_processor.clj:470)"
  "driver.sql.query_processor$fn__69112.invokeStatic(query_processor.clj:742)"
  "driver.sql.query_processor$fn__69112.invoke(query_processor.clj:740)"
  "driver.sql.query_processor$fn__69124.invokeStatic(query_processor.clj:750)"
  "driver.sql.query_processor$fn__69124.invoke(query_processor.clj:748)"
  "driver.sql.query_processor$apply_top_level_clauses$fn__69164.invoke(query_processor.clj:838)"
  "driver.sql.query_processor$apply_top_level_clauses.invokeStatic(query_processor.clj:836)"
  "driver.sql.query_processor$apply_top_level_clauses.invoke(query_processor.clj:832)"
  "driver.sql.query_processor$apply_clauses.invokeStatic(query_processor.clj:913)"
  "driver.sql.query_processor$apply_clauses.invoke(query_processor.clj:898)"
  "driver.sql.query_processor$fn__69241$build_honeysql_form__69246$fn__69250.invoke(query_processor.clj:918)"
  "driver.sql.query_processor$fn__69241$build_honeysql_form__69246.invoke(query_processor.clj:915)"
  "driver.sql.query_processor$fn__69070.invokeStatic(query_processor.clj:704)"
  "driver.sql.query_processor$fn__69070.invoke(query_processor.clj:696)"
  "driver.sql.query_processor$fn__69082$join__GT_honeysql__69075__69087$fn__69091.invoke(query_processor.clj:719)"
  "driver.sql.query_processor$fn__69082$join__GT_honeysql__69075__69087.invoke(query_processor.clj:717)"
  "driver.sql.query_processor$fn__69104$fn__69107.invoke(query_processor.clj:733)"
  "driver.sql.query_processor$fn__69104.invokeStatic(query_processor.clj:731)"
  "driver.sql.query_processor$fn__69104.invoke(query_processor.clj:729)"
  "driver.sql.query_processor$apply_top_level_clauses$fn__69164.invoke(query_processor.clj:838)"
  "driver.sql.query_processor$apply_top_level_clauses.invokeStatic(query_processor.clj:836)"
  "driver.sql.query_processor$apply_top_level_clauses.invoke(query_processor.clj:832)"
  "driver.sql.query_processor$apply_clauses.invokeStatic(query_processor.clj:913)"
  "driver.sql.query_processor$apply_clauses.invoke(query_processor.clj:898)"
  "driver.sql.query_processor$fn__69241$build_honeysql_form__69246$fn__69250.invoke(query_processor.clj:918)"
  "driver.sql.query_processor$fn__69241$build_honeysql_form__69246.invoke(query_processor.clj:915)"
  "driver.sql.query_processor$mbql__GT_honeysql.invokeStatic(query_processor.clj:929)"
  "driver.sql.query_processor$mbql__GT_honeysql.invoke(query_processor.clj:927)"
  "driver.sql.query_processor$mbql__GT_native.invokeStatic(query_processor.clj:935)"
  "driver.sql.query_processor$mbql__GT_native.invoke(query_processor.clj:931)"
  "driver.sql$fn__78087.invokeStatic(sql.clj:40)"
  "driver.sql$fn__78087.invoke(sql.clj:38)"
  "query_processor.middleware.mbql_to_native$query__GT_native_form.invokeStatic(mbql_to_native.clj:15)"
  "query_processor.middleware.mbql_to_native$query__GT_native_form.invoke(mbql_to_native.clj:10)"
  "query_processor.middleware.mbql_to_native$mbql__GT_native$fn__47424.invoke(mbql_to_native.clj:23)"
  "query_processor.middleware.check_features$check_features$fn__46699.invoke(check_features.clj:42)"
  "query_processor.middleware.optimize_datetime_filters$optimize_datetime_filters$fn__47589.invoke(optimize_datetime_filters.clj:133)"
  "query_processor.middleware.auto_parse_filter_values$auto_parse_filter_values$fn__45500.invoke(auto_parse_filter_values.clj:44)"
  "query_processor.middleware.wrap_value_literals$wrap_value_literals$fn__49155.invoke(wrap_value_literals.clj:149)"
  "query_processor.middleware.annotate$add_column_info$fn__45262.invoke(annotate.clj:575)"
  "query_processor.middleware.permissions$check_query_permissions$fn__46574.invoke(permissions.clj:70)"
  "query_processor.middleware.pre_alias_aggregations$pre_alias_aggregations$fn__48107.invoke(pre_alias_aggregations.clj:40)"
  "query_processor.middleware.cumulative_aggregations$handle_cumulative_aggregations$fn__46772.invoke(cumulative_aggregations.clj:61)"
  "query_processor.middleware.resolve_joined_fields$resolve_joined_fields$fn__48344.invoke(resolve_joined_fields.clj:36)"
  "query_processor.middleware.resolve_joins$resolve_joins$fn__48663.invoke(resolve_joins.clj:183)"
  "query_processor.middleware.add_implicit_joins$add_implicit_joins$fn__40987.invoke(add_implicit_joins.clj:254)"
  "query_processor.middleware.large_int_id$convert_id_to_string$fn__47385.invoke(large_int_id.clj:44)"
  "query_processor.middleware.limit$limit$fn__47410.invoke(limit.clj:38)"
  "query_processor.middleware.format_rows$format_rows$fn__47365.invoke(format_rows.clj:84)"
  "query_processor.middleware.desugar$desugar$fn__46838.invoke(desugar.clj:22)"
  "query_processor.middleware.binning$update_binning_strategy$fn__45859.invoke(binning.clj:229)"
  "query_processor.middleware.resolve_fields$resolve_fields$fn__46376.invoke(resolve_fields.clj:24)"
  "query_processor.middleware.add_dimension_projections$add_remapping$fn__40531.invoke(add_dimension_projections.clj:318)"
  "query_processor.middleware.add_implicit_clauses$add_implicit_clauses$fn__40738.invoke(add_implicit_clauses.clj:141)"
  "query_processor.middleware.add_source_metadata$add_source_metadata_for_source_queries$fn__41136.invoke(add_source_metadata.clj:105)"
  "query_processor.middleware.reconcile_breakout_and_order_by_bucketing$reconcile_breakout_and_order_by_bucketing$fn__48304.invoke(reconcile_breakout_and_order_by_bucketing.clj:98)"
  "query_processor.middleware.auto_bucket_datetimes$auto_bucket_datetimes$fn__45447.invoke(auto_bucket_datetimes.clj:125)"
  "query_processor.middleware.resolve_source_table$resolve_source_tables$fn__46423.invoke(resolve_source_table.clj:46)"
  "query_processor.middleware.parameters$substitute_parameters$fn__48089.invoke(parameters.clj:114)"
  "query_processor.middleware.resolve_referenced$resolve_referenced_card_resources$fn__46475.invoke(resolve_referenced.clj:80)"
  "query_processor.middleware.expand_macros$expand_macros$fn__47094.invoke(expand_macros.clj:158)"
  "query_processor.middleware.add_timezone_info$add_timezone_info$fn__41167.invoke(add_timezone_info.clj:15)"
  "query_processor.middleware.splice_params_in_response$splice_params_in_response$fn__49025.invoke(splice_params_in_response.clj:32)"
  "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__48315$fn__48319.invoke(resolve_database_and_driver.clj:33)"
  "driver$do_with_driver.invokeStatic(driver.clj:61)"
  "driver$do_with_driver.invoke(driver.clj:57)"
  "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__48315.invoke(resolve_database_and_driver.clj:27)"
  "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__47312.invoke(fetch_source_query.clj:267)"
  "query_processor.middleware.store$initialize_store$fn__49034$fn__49035.invoke(store.clj:11)"
  "query_processor.store$do_with_store.invokeStatic(store.clj:46)"
  "query_processor.store$do_with_store.invoke(store.clj:40)"
  "query_processor.middleware.store$initialize_store$fn__49034.invoke(store.clj:10)"
  "query_processor.middleware.cache$maybe_return_cached_results$fn__46352.invoke(cache.clj:214)"
  "query_processor.middleware.validate$validate_query$fn__49043.invoke(validate.clj:10)"
  "query_processor.middleware.normalize_query$normalize$fn__47437.invoke(normalize_query.clj:22)"
  "query_processor.middleware.add_rows_truncated$add_rows_truncated$fn__41005.invoke(add_rows_truncated.clj:36)"
  "query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__49010.invoke(results_metadata.clj:147)"
  "query_processor.middleware.constraints$add_default_userland_constraints$fn__46715.invoke(constraints.clj:42)"
  "query_processor.middleware.process_userland_query$process_userland_query$fn__48178.invoke(process_userland_query.clj:136)"
  "query_processor.middleware.catch_exceptions$catch_exceptions$fn__46658.invoke(catch_exceptions.clj:174)"
  "query_processor.reducible$async_qp$qp_STAR___39794$thunk__39795.invoke(reducible.clj:103)"
  "query_processor.reducible$async_qp$qp_STAR___39794.invoke(reducible.clj:109)"
  "query_processor.reducible$sync_qp$qp_STAR___39803$fn__39806.invoke(reducible.clj:135)"
  "query_processor.reducible$sync_qp$qp_STAR___39803.invoke(reducible.clj:134)"
  "query_processor$process_userland_query.invokeStatic(query_processor.clj:237)"
  "query_processor$process_userland_query.doInvoke(query_processor.clj:233)"
  "query_processor$fn__49201$process_query_and_save_execution_BANG___49210$fn__49213.invoke(query_processor.clj:249)"
  "query_processor$fn__49201$process_query_and_save_execution_BANG___49210.invoke(query_processor.clj:241)"
  "query_processor$fn__49245$process_query_and_save_with_max_results_constraints_BANG___49254$fn__49257.invoke(query_processor.clj:261)"
  "query_processor$fn__49245$process_query_and_save_with_max_results_constraints_BANG___49254.invoke(query_processor.clj:254)"
  "api.dataset$fn__52570$fn__52573.invoke(dataset.clj:59)"
  "query_processor.streaming$streaming_response_STAR_$fn__37098$fn__37099.invoke(streaming.clj:73)"
  "query_processor.streaming$streaming_response_STAR_$fn__37098.invoke(streaming.clj:72)"
  "async.streaming_response$do_f_STAR_.invokeStatic(streaming_response.clj:66)"
  "async.streaming_response$do_f_STAR_.invoke(streaming_response.clj:64)"
  "async.streaming_response$do_f_async$fn__24643.invoke(streaming_response.clj:85)"],
 :context :ad-hoc,
 :error "No aggregation at index: 0",
 :row_count 0,
 :running_time 0,
 :preprocessed
 {:type :query,
  :query
  {:source-table 10,
   :joins
   [{:strategy :left-join,
     :condition [:= [:field-id 107] [:joined-field "Question 351" [:field-literal "CATEGORY" :type/Text]]],
     :alias "Question 351",
     :source-query {:source-table 10, :aggregation [[:aggregation-options [:count] {:name "count"}]], :breakout [[:field-id 107]], :order-by [[:asc [:aggregation 0]]]},
     :source-metadata
     [{:base_type :type/Text,
       :display_name "Category",
       :name "CATEGORY",
       :description "The type of product, valid values include: Doohicky, Gadget, Gizmo and Widget",
       :special_type :type/Category,
       :fingerprint {:global {:distinct-count 4, :nil% 0}, :type {:type/Text {:percent-json 0, :percent-url 0, :percent-email 0, :percent-state 0, :average-length 6.375}}}}
      {:base_type :type/BigInteger,
       :display_name "Count",
       :name "count",
       :special_type :type/Quantity,
       :fingerprint {:global {:distinct-count 4, :nil% 0}, :type {:type/Number {:min 42, :q1 46.5, :q3 53.5, :max 54, :sd 5.477225575051661, :avg 50}}}}]}],
   :fields
   [[:field-id 105]
    [:field-id 102]
    [:field-id 106]
    [:field-id 107]
    [:field-id 109]
    [:field-id 104]
    [:field-id 103]
    [:datetime-field [:field-id 108] :default]
    [:joined-field "Question 351" [:field-literal "CATEGORY" :type/Text]]
    [:joined-field "Question 351" [:field-literal "count" :type/BigInteger]]],
   :limit 2000},
  :database 4,
  :middleware {:js-int-to-string? true, :add-default-userland-constraints? true},
  :info
  {:executed-by 1,
   :context :ad-hoc,
   :nested? false,
   :query-hash [-123, -64, 57, -14, -86, -92, -55, -110, 104, 95, -43, 122, 45, 69, -73, -36, -28, -19, 26, -46, 40, -120, 114, -46, -80, 97, 64, -63, -12, -114, 65, -14]},
  :constraints {:max-results 10000, :max-results-bare-rows 2000}},
 :data {:rows [], :cols []}}

2020-11-10 14:30:55,684 DEBUG middleware.log :: POST /api/dataset 202 [ASYNC: completed] 349.5 ms (20 DB calls) App DB connections: 0/10 Jetty threads: 3/50 (3 idle, 0 queued) (113 total active threads) Queries in flight: 0 (0 queued)
```

</details>

**Information about your Metabase Installation:**
Tested Metabase 0.34.3 thru 0.37.0.2

**Additional context**
Related to #8283 and #6239

**Comment by camsaul** [comment]
Based on the info above, this is the failing query:

```clj
{:type "query",
 :query
 {:source-table 10,
  :joins
  [{:fields "all",
    :source-query
    {:source-table 10,
     :aggregation [[:count]],
     :breakout [[:field-id 107]],
     :order-by [[:asc [:aggregation 0]]]},
    :condition
    ["="
     ["field-id" 107]
     ["joined-field"
      "Question 351"
      ["field-literal" "CATEGORY" "type/Text"]]],
    :alias "Question 351"}]},
 :database 4,
 :parameters [],
 :middleware
 {:js-int-to-string? true, :add-default-userland-constraints? true}}
```

It looks correct, I don't see any reason why it shouldn't work. It seems like it's not handling the `:aggregation` reference correctly because it's inside of a source query inside of a join. We use a dynamic variable `*query*` to track the current query level as we're compiling stuff... we probably need to make sure this gets bound when compiling joins, otherwise we'll look in the wrong level to try to find the aggregation

**Comment by camsaul** [comment]
Fixed by #19384
